lib/internal/http2/compat.js

```diff
@@ -330,45 +306,97 @@ class Http2ServerResponse extends Stream {
     state.statusCode = code;
   }
 
+  addTrailers(headers) {
+    for (var name of Object.keys(headers)) {
+      name = String(name).trim().toLowerCase();
+      var value = headers[name];
+      assertValidHeader(name, value);
+      var trailers = this[kTrailers];
+      if (!trailers)
+        trailers = this[kTrailers] = Object.create(null);
+      trailers[name] = String(value);
+    }
+  }
+
+  getHeader(name) {
+    var headers = this[kHeaders];
+    if (!headers)
+      return;
+    name = String(name).trim().toLowerCase();
+    return headers[name];
+  }
+
+  getHeaderNames() {
+    var headers = this[kHeaders];
+    if (!headers)
+      return [];
+    return Object.keys(headers);
+  }
+
+  getHeaders() {
+    var headers = this[kHeaders];
+    return Object.assign({}, headers);
+  }
+
+  hasHeader(name) {
+    var headers = this[kHeaders];
+    if (!headers)
+      return false;
+    name = String(name).trim().toLowerCase();
+    return Object.prototype.hasOwnProperty.call(headers, name);
+  }
+
+  removeHeader(name) {
+    var headers = this[kHeaders];
+    if (!headers)
+      return;
+    name = String(name).trim().toLowerCase();
+    delete headers[name];
+  }
+
   setHeader(name, value) {
     var state = this[kState];
+    if (state.headersSent)
+      throw new Error('Cannot set status after the HTTP message has been sent');
+    name = String(name).trim().toLowerCase();
+    assertValidHeader(name, value);
     var headers = this[kHeaders];
-    if (state.headersSent) {
-      throw new Error(
-        'Cannot set headers after the HTTP message has been sent');
-    }
-    setHeader(headers, name, value);
-    state.headerCount++;
-    return this;
+    if (!headers)
+      headers = this[kHeaders] = Object.create(null);
+    headers[name] = String(value);
   }
 
-  setTrailer(name, value) {
+  flushHeaders() {
     var state = this[kState];
-    var trailers = this[kTrailers];
-    if (state.trailersSent) {
-      throw new Error(
-        'Cannot set trailers after the HTTP message has been sent');
-    }
-    setHeader(trailers, name, value);
-    state.trailerCount++;
-    return this;
+    if (state.headersSent)
+      return;
+    var statusCode = this[kState].statusCode;
+    this.writeHead(statusCode);
   }
 
   writeHead(statusCode, headers) {
-    var keys = Object.keys(headers);
-    var key = '';
-    for (var i = 0; i < keys.length; i++) {
-      key = keys[i];
-      this.setHeader(key, headers[key]);
+    var state = this[kState];
+    if (state.headersSent)
+      return;
+    if (headers) {
+      var keys = Object.keys(headers);
+      var key = '';
+      for (var i = 0; i < keys.length; i++) {
+        key = keys[i];
+        this.setHeader(key, headers[key]);
+      }
     }
     this.statusCode = statusCode;
     this[kBeginSend]();
-    return this;
+  }
+
+  writeHeader(statusCode, headers) {
+    return this.writeHead.apply(this, arguments);
```

Comment:
TIL `apply`/`call` are about [half the speed](https://jsperf.com/function-calls-direct-vs-apply-vs-call-vs-bind/6) of a direct call on Chrome.

Saw writeHeaders in `_http_server.js` being forwarded to writeHead [exactly this way using `apply`](https://github.com/nodejs/http2/blob/master/lib/_http_server.js#L232-L234).

Looks like [the writeHeaders API has been deprecated](https://github.com/nodejs/node/commit/fb71ba49212dd21c0404892b4bddc37437bd5397) and dropped from the main nodejs/node repo.

Will remove it.